crates/shim-sgx/src/handler/mod.rs

```diff
@@ -230,9 +252,33 @@ impl guest::Handler for Handler<'_> {
     fn munmap(
         &mut self,
         _platform: &impl Platform,
-        _addr: NonNull<c_void>,
-        _length: c_size_t,
+        addr: NonNull<c_void>,
+        length: c_size_t,
     ) -> sallyport::Result<()> {
+        let addr = addr.as_ptr() as usize;
+        let length = (length + 0xfff) & !0xfff;
+
+        // According to the man-page: munmap (3)
+        if addr & 0xfff != 0 || length == 0 {
+            return Err(EINVAL);
+        }
+
+        let region = Region::new(
+            Address::<usize, Page>::new(addr),
+            Address::<usize, Page>::new(addr + length),
+        );
+
+        // According to the man-page: munmap (3)
+        if HEAP.read().contains(region) == None {
+            return Err(EINVAL);
+        }
+
+        self.trim_sgx_pages(NonNull::new(addr as *mut _).unwrap(), length)?;
+        self.accept_munmap(region);
+        self.remove_sgx_pages(NonNull::new(addr as *mut _).unwrap(), length)?;
+
+        HEAP.write().munmap(region);
+
```

Comment:
I was not sure whether I can make up-calls with these spinlocks held. @npmccallum, @haraldh? So I'm aware that this is racy but need the feedback whether I can take the lock and make an up-call.

Reply:
Works so I close this.